Fix lsim

cwrowley opened the pull request:
This pull request provides a rewrite of `signal.lsim`, and fixes #2652 and #1104.
The new implementation fixes several known problems, in particular for systems with a zero eigenvalue, or with repeated eigenvalues.  It is also much faster than `lsim2`, which calls a generic ODE integrator.  See the commit messages for more details.

rgommers commented:
This looks promising, thanks @cwrowley.

rgommers commented:
First looking at the test changes, I see that ``test_05`` and ``test_06`` for ``TestStep2`` are almost the same as the ones they overwrite in the base test class - only difference is ``atol=1e-10, rtol=1e-8``. If that's necessary, adding a comment to explain it would be good.

Other than that ``step/step2`` and ``impulse/impulse2`` are the same, so ``step2, impulse2`` could be deprecated it looks like.

rgommers commented:
Since people (like me) usually don't read individual commit messages, I'll copy one here:

    The previous implementation of lsim did not work for systems with a pole
    at the origin, or with repeated poles.  This new version fixes all of
    the (known) issues with lsim.

    The new implementation uses the matrix exponential, like the old
    implementation, but computes it using linalg.expm instead of by
    diagonalization.  This solves the problem with repeated poles.  In
    addition, the new implementation does not require inv(A) to compute the
    forcing term, and it no longer fails for systems with a pole at the
    origin.  Details of the algorithm are documented in the comments.

    Note that there is already an lsim2 routine that fixes the above
    problems.  However, it calls a generic ODE integrator, and can be
    hundreds of times slower than the implementation here.

    An important change with the new implementation is that the time vector
    must now be evenly spaced.  Note that this same requirement is in the
    MATLAB version of lsim.  One could modify this to allow for non-uniform
    time samples, but this would slow down the algorithm, because the matrix
    exponential would need to be recomputed whenever the timestep changes.
    In addition, small, machine-precision-level differences in timestep
    might trigger the matrix exponential to be recomputed when it is not
    necessary.  Therefore, the implementation here checks whether timesteps
    are approximately uniform using np.allclose(), and gives an error if
    they are not.

rgommers commented:
I wouldn't expect the equal spacing requirement to be an issue, but technically it's a backwards compatibility break so it would be good to ask on the mailing list if anyone relies on the current behavior (and if so, can they work around this PR).

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
-
-    yout = (squeeze(dot(U, transpose(sys.D))) +
-            squeeze(dot(xout, transpose(sys.C))))
+    if not np.allclose(T[1:] - T[:-1], dt):
```
This may not work perfectly for time arrays with units very far from seconds (``atol=1e-8`` is not scale invariant), maybe divide ``T`` by ``dt`` and compare to ``1``?

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -790,16 +835,7 @@ def impulse(system, X0=None, T=None, N=None):
-        eA = _cast_to_array_dtype(eA, h)
-        h[k] = squeeze(dot(dot(C, eA), B))
-
+    _t, h, _x = lsim(sys, 0., T, X, interp=0)
```
variables you don't use you can simply name ``_``.

rgommers replied:
Can you make this ``interp=False`` and change the description of ``interp`` in the ``lsim`` docstring to be a ``bool``?

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+
+    no_input = (U is None
+                or (isinstance(U, (int, float)) and U == 0.)
+                or np.allclose(U, 0))
```
This ``allclose`` usage looks wrong:

    In [1]: np.allclose([0, 1e-9, 2e-9], 0)
    Out[1]: True

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+    n_states = A.shape[0]
+    n_inputs = B.shape[1]
+
+    n_steps = len(T)
```
``T.size``

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
-    if sU[1] != sys.inputs:
-        raise ValueError("System does not define that many inputs.")
 
+    A, B, C, D = (np.asarray(M) for M in (sys.A, sys.B, sys.C, sys.D))
```
Maybe a bit more idiomatic: ``map(np.asarray, (sys.A, sys.B, sys.C, sys.D))``

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+
+    # Nonzero input
+    U = atleast_1d(U)
+    if len(U.shape) == 1:
```
``U.ndim == 1``

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+    U = atleast_1d(U)
+    if len(U.shape) == 1:
+        U = U.reshape((U.shape[0], 1))
+    sU = U.shape
```
Better for readability to just use ``U.shape`` where needed.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+    # Nonzero input
+    U = atleast_1d(U)
+    if len(U.shape) == 1:
+        U = U.reshape((U.shape[0], 1))
```
maybe ``U = U[:, np.newaxis]``

rgommers commented:
Just to make sure I understand usage of ``lti_nowarn`` in the tests: these are valid ABCD inputs, however because we're converting to transfer function representation we get warnings due to the tf format not being able to express these systems? So we can fix that only after we get rid of all those conversions (not soon probably).

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+            xout[i] = (dot(xout[i-1], Ad) + dot(U[i-1], Bd0) +
+                       dot(U[i], Bd1))
+    yout = (squeeze(dot(xout, transpose(C))) +
+            squeeze(dot(U, transpose(D))))
```
This fits on 1 line

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+        Bd0 = expMT[n_states:n_states + n_inputs, :n_states] - Bd1
+        for i in xrange(1, n_steps):
+            xout[i] = (dot(xout[i-1], Ad) + dot(U[i-1], Bd0) +
+                       dot(U[i], Bd1))
```
fits on 1 line as well

rgommers replied:
and style nitpick: please insert a blank line below and between consecutive if-statements in a few places above

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
+        # Solution is
+        #   [ x(dt) ]       [ A*dt   B*dt ] [ x0 ]
+        #   [ u(dt) ] = exp [  0     0    ] [ u0 ]
+        M = np.bmat([[A * dt, B * dt],
```
Please don't use this function or ``np.matrix`` instances. We're trying to keep Scipy code free of ``np.matrix`` as far as possible. 

This should be equivalent to something like ``np.vstack([np.hstack([A * dt, B * dt]), np.zeros(...)])``

rgommers replied:
And same for second ``bmat`` usage below

rgommers commented:
@cwrowley do you have a reference for this algorithm, or did you write it from scratch? If the former, would be good to add it to the docstring.

Adding an example to the docstring would also be nice. Should be easy to convert a test to an example and add a plot with the ``lsim`` result.

WarrenWeckesser commented:
As the author of the work-around `step2()`, let me say "Thanks!" for working on this!  Deprecating `lsim2()`, `step2()` and `impulse2()` in favor of improved `lsim()`, `step()` and `impulse()` functions is a nice improvement.

I won't have time to help out with the code review (unfortunately), but this change caught my eye:

> An important change with the new implementation is that the time vector must now be evenly spaced.

@rgommers wrote:

> I wouldn't expect the equal spacing requirement to be an issue,

I'm not so sure about that.  It isn't too unusual to use non-uniform spacing (e.g. sometimes an expanding sequence such as `t = np.logspace(-2, 2, 13)` can be useful).  I don't know how common it is to use `lsim()` with irregular spacing, but it has always worked.  I recommend finding a way to maintain backwards compatibility here.

cwrowley commented:
I think I addressed all the points raised above by @rgommers.
> @cwrowley do you have a reference for this algorithm, or did you write it from scratch? If the former, would be good to add it to the docstring.

I wrote it from scratch, and I'm not aware of anything to reference.

@rgommers wrote:
> Just to make sure I understand usage of lti_nowarn in the tests: these are valid ABCD inputs, however because we're converting to transfer function representation we get warnings due to the tf format not being able to express these systems? So we can fix that only after we get rid of all those conversions (not soon probably).

These are valid ABCD inputs, and the transfer functions are also perfectly valid (e.g., 1/s^2 for the double integrator), but when the `lti` constructor converts it to a transfer function, it complains about "Badly conditioned filter coefficients (numerator)."  So yes, I think it's correct that we can fix this only after getting rid of all these conversions.

About the spacing issue, while I agree that non-uniform spacing is often useful in things like a frequency response, at least in my experience, it is rarely if ever useful for time responses.  Note that Matlab also requires uniform timesteps in their version of `lsim`.  Perhaps there is a case where it comes up, but in 15 years of teaching control systems, I have never once wished Matlab could accommodate nonuniform timesteps.  If one did have a case where the timestep were changing at each step, then I think this would be the wrong algorithm to use: it would probably be more efficient to call a generic ODE integrator, as done in `lsim2`, rather than recompute the matrix exponential at each step.  And if the timestep were changing just in a few places, a workaround could be to call `lsim` multiple times, once for each value of the timestep.

@rgommers wrote:
> it would be good to ask on the mailing list if anyone relies on the current behavior (and if so, can they work around this PR).

I just posted a message to the scipy-users mailing list, so we'll see if anybody has any thoughts.

moorepants commented:
Not that I have a use case for non-uniform time, but would it be too complex to check for non-uniform time and route to the less efficient routine? If that routine is already in lsim2, why not reroute to it?

cwrowley commented:
@moorepants, in principle that would be fine, I suppose.  One issue is the interface: should one then be able to pass additional keyword arguments through to `scipy.integrate.odeint` as in `lsim2`? This can be necessary with `lsim2` in order to specify suitable tolerances for `odeint`.  (By contrast, the method used here is "exact", at least to the precision of `scipy.linalg.expm`, which is probably machine precision, so these tolerances are unnecessary.)  In my view, this could be potentially confusing to the user, to be able to specify keyword arguments that are unused except in a highly unusual case of nonuniform timesteps.

Another issue to consider is surprising behavior.  If we use two completely different algorithms, depending on the input, this could lead to surprising behavior for the user: for one type of input (uniform steps), the routine would be fast and accurate, while for another type of input (nonuniform steps), it would be hundreds of times slower and less accurate. True, with the version in this PR, the second case would throw an exception, but in my view that is better, since at least there would be an explanation for the behavior.  (And that is also consistent with what Matlab and Octave do.)

So my preference would be for the simpler, cleaner interface that does not give surprising behavior.  I recognize it is a backwards-compatibility break, but in my view it was a design mistake to allow this behavior in the original version of `lsim`, and I think this would be a good chance to fix it.  But I am happy to change this if that is what others prefer.

moorepants commented:
Does SciPy have a deprecation policy/guideline? Seems like you would need to support the previous behavior and issue a warning if non-uniform time is supplied for at least one release cycle before removing the functionality if that is the ultimate goal.

I generally favor avoiding API breaks to fixing design mistakes, especially on widely used packages, but that's of course my opinion. I won't hold this up either way you go.

argriffing commented:
> the method used here is "exact", at least to the precision of scipy.linalg.expm, which is probably machine precision

The `scipy.linalg.expm` is currently intended to be an implementation of http://epubs.siam.org/doi/abs/10.1137/09074721X which is not necessarily accurate to machine precision in the sense of having guarantees like http://www.gnu.org/software/libc/manual/html_node/Errors-in-Math-Functions.html#Errors-in-Math-Functions or http://fredrikj.net/arb/acb_mat.html#c.acb_mat_exp.

argriffing commented:
No one has responded on the mailing list or on here in the last week, so I'll add something.

@cwrowley If the less-than-perfect precision of `expm` is not a deal-breaker then let's do this! But it needs to keep backwards compatibility with scipy, not with Matlab/Octave. Some benchmarks should be added if a substantial part of the PR's justification is to improve speed, especially if there is a loss of precision or a loss of control over precision. Also I'm not sure about complex-valued inputs (like imaginary numbers). Maybe they aren't relevant but the new tests don't include them.

Just to give some historical background, the scipy signal module is in a weird position where the function names and behavior seem like they should be clones of Matlab, but this is misleading. It's a different situation from Octave which is directly a Matlab clone, and in fact scipy seems to be moving away from the Matlab names when new functions are added. Regarding this PR another option could be to add the new implementation as `lsim3` but I don't suppose this would be a popular idea.

Personally I don't know anything about control theory so I'll need some help if I try to merge this into scipy. I have some experience with statistical inference in finite-state continuous-time Markov processes which appears to use some of the same math, but I have no experience with dynamic control, so linear algebra notation helps me more than control/signal jargon. Although it may be out of scope for this PR, there could also be opportunities to use `expm_multiply` somehow, which theoretically implements matrix-exponential-vector-products over equally spaced time intervals. This would also make better use of the structural sparsity of the input matrix, or on the other hand maybe the extra complication of the control vector would rule out the use of this function. It could also be possible to revive an old expokit PR https://github.com/scipy/scipy/pull/354 if this implements a more appropriate exponential integrator; the expokit author Professor Roger Sidje appears to have given license permission according to http://mail.scipy.org/pipermail/scipy-user/2012-November/033600.html.

moorepants commented:
What about giving this function a proper Python style name and keeping it separate from the previous implementations since it doesn't actually reproduce their results?

Name suggestions:
- `simulate_linear_system()`
- `linear_sim()`
- 'integrate_linear_system()`
- `sim_lin_sys()`
- `con_simulation()`
- `simulate()`
- etc

Of course, the `lsim` and `lsim2` may eventually be deprecated in favor of this.

cwrowley commented:
@argriffing I don't think falling short of machine precision is a problem at all.  I am sure that `scipy.linalg.expm` is more accurate than the way the matrix exponential is computed in the current implementation of `lsim`, and it is also more accurate than the generic ODE integrator used in `lsim2` (that is why `lsim2` requires a specialized version of one of the tests, to tighten up the tolerances).

Regarding benchmarks, I am happy to add a benchmark (if I can figure out how), but the main reason for this PR is not to improve speed, it's to fix a routine that gives incorrect results.  The new algorithm isn't fundamentally different from the existing algorithm--it is still based on the matrix exponential--but it is improved so that it works for a wider class of systems (e.g., repeated eigenvalues, and systems with a zero eigenvalue, such as an integrator).

My understanding (although @WarrenWeckesser would certainly know better) is that `lsim2` was originally added as a temporary workaround, to fix these problems with `lsim`, and the plan was for it eventually to be removed.  So I would not recommend adding a third version `lsim3`.

Also, I did not mean to suggest that we should strive to mimic Matlab--my remark was just meant to support my claim that non-uniform timesteps are not necessary, and that allowing them is actually a design flaw (in my opinion).

@moorepants I also would prefer a more descriptive, pythonic name (e.g., the corresponding routine in the [python-control](https://github.com/python-control/python-control) package is called `forced_reponse`), but I think that in any case, the existing `lsim` routine needs to be fixed or removed.

If maintaining backward compatibility is important, then how about doing what @moorepants suggested and making support for nonuniform timesteps deprecated?  So if nonuniform timesteps are used, then we could call through to `lsim2` (with default tolerances), and issue a DeprecationWarning that results may be slow and/or inaccurate, and support for nonuniform timesteps may be removed in the future.  That should be simple to add, and would still give us a way forward towards eliminating the redundant routines `lsim2`, `step2`, etc.  Does that sound reasonable?

moorepants commented:
Sounds good to me.

argriffing commented:
Added the benchmark as https://github.com/scipy/scipy/pull/4766.

argriffing commented:
FWIW here's code to reproduce the first matlab lsim docs example.

```python
import numpy as np
from scipy.signal import lti, lsim
import matplotlib.pyplot as plt

def main():
    systems = (lti([2, 5, 1], [1, 2, 3]), lti([1, -1], [1, 1, 5]))
    rep = [0.0]*20 + [1.0]*20
    u = (rep + rep + rep)[:101]
    t = np.linspace(0, 10, num=len(u))
    f, axarr = plt.subplots(2, sharex=True)
    for s, ax in zip(systems, axarr):
        tout, y, x = lsim(s, u, t, interp=0)
        ax.plot(t, y)
        ax.plot(t, u)
    plt.show()

main()
```

cwrowley commented:
Should `step2`, `impulse2`, and `lsim2` then be deprecated?  I am not sure about the right way to do this, but perhaps I should add a `DeprecationWarning` to those routines, as in commit 822d7e6?

rgommers commented:
Passing through to ``lsim2`` looks right to me. And I agree with the comment of @cwrowley on more Pythonic names - would have been the right choice when starting from scratch, but now that we have ``lsim`` let's stick to it.

@moorepants Scipy deprecation policy is to at least have one release with a warning before removing something. The exception is if something is clearly buggy and keeping it around does more harm than good.

rgommers commented:
@cwrowley may be good to wait one or two releases with deprecating ``step2/lsim2/impulse2``. Otherwise everyone who wants his code to work with the current and previous release has to do something like:

    if scipy.__version__ < '0.16.0'  # this is not even robust, needs more complex parsing
        lsim2(...)
    else:
        lsim(...)

argriffing commented:
Another interface option could be to just use [`linspace`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linspace.html)-like args (that is, `start, stop, num=50, endpoint=True, retstep=False`) instead of `T`, together with a new function name. I'm not necessarily proposing this idea, just raising it as a possibility.

argriffing commented:
This PR changes `interp=1` to `interp=True`. Would higher dimensional interpolation be possible in the future, or is there an obstruction preventing higher order interpolation?

cwrowley commented:
I don't know if there is a fundamental obstruction to higher-order interpolation, but I don't see a nice way to do it off the top of my head.  The reason I changed that is in response to the suggestion by @rgommers in the line notes above:
> Can you make this interp=False and change the description of interp in the lsim docstring to be a bool?

rgommers commented:
The current ``interp=1`` seems more like a misplaced use of ints as booleans (the code only has ``if interp:``), which was common in older scipy code. I checked matlab, and it uses strings (``'zoh'`` and ``''foh'``) which also isn't super clear. Looking at the current code, I don't see how you'd add higher-order interpolation without making the code much slower. Also not sure that there's a usecase.

ev-br commented:
Is https://github.com/scipy/scipy/pull/2862 anyhow relevant?

argriffing commented:
> I checked matlab, and it uses strings ('zoh' and ''foh') which also isn't super clear. Looking at the current code, I don't see how you'd add higher-order interpolation without making the code much slower. Also not sure that there's a usecase.

According to the internet zoh and foh are zero and first order holds respectively, and second order holds are a thing. But this doesn't necessarily have to interfere with this PR.

argriffing commented:
@ev-br The test in that PR (still) fails at the `lti(A, B, C, D)` step before the `lsim` changes come into play.

argriffing commented:
> Is #2862 anyhow relevant?

To try speeding up this PR I've opened a new issue https://github.com/scipy/scipy/issues/4794 which isolates one of the two problems from https://github.com/scipy/scipy/pull/2862, but these problems seem to be out of scope for this PR in my opinion.

The first problem is that the `lti` constructor cannot deal with some `(A, B, C, D)` tuples which work perfectly well with `lsim`. You could work around this by working directly with a state space system tuple instead of through a lti object:
```python
    if isinstance(system, lti):
        sys = system
    elif len(system) == 4:
        sys = namedtuple('ss', ['A', 'B', 'C', 'D'])(*system)
    else:
        sys = lti(*system)
```

The second problem seems to be due to oversqueezing.
```python
return T, squeeze(yout), squeeze(xout)
```
This transforms `xout` from a column vector with shape `(3, 1)` to a 1d vector with shape `(3, )` which according to https://github.com/scipy/scipy/pull/2862 is bad for some reason that I don't understand.

@rgommers has had the opinion in https://github.com/scipy/scipy/pull/2862#issuecomment-25744572 that `Something should be added to the docs about MIMO systems I think. And then it should either work for all functions or for none of them.` which suggests these changes would be out of scope for this PR.

rgommers commented:
@argriffing I agree that MIMO systems are out of scope for this PR. This is pretty much done and should be merged soon.

rgommers commented:
OK, backwards compat issue fixed, added example looks good, so in this goes.

rgommers commented:
Merged, thanks @cwrowley!
